util/metrics/tdigest_sample.go

```diff
@@ -0,0 +1,216 @@
+	for ; i < len(sample.buckets); i++ {
+		if sample.buckets[i].earliest.After(cutoff) {
+			break
+		}
+	}
+	sample.buckets = sample.buckets[i:]
+}
+
+func (sample *TDigestSample) Update(value int64) {
+	sample.mutex.Lock()
+	defer sample.mutex.Unlock()
+
+	sample.count++
+	now := sample.clock.Now()
+	if now.Sub(sample.cur.earliest) >= bucketWidth {
+		sample.buckets = append(sample.buckets, sample.cur)
+		sample.drop()
+		sample.cur = sample.newBucket()
+	}
+
+	sample.cur.values += value
+	sample.cur.count++
+	sample.cur.digest.Add(float64(value), 1.0)
+}
+
+func (sample *TDigestSample) Clear() {
+	sample.mutex.Lock()
+	defer sample.mutex.Unlock()
+
+	sample.count = 0
+	sample.buckets = nil
+	sample.cur = sample.newBucket()
+}
+
+func (sample *TDigestSample) Count() int64 {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	return sample.count
+}
+
+func (sample *TDigestSample) Dropped() int64 {
+	return 0
+}
+
+func (sample *TDigestSample) Size() int {
+	panic("not implemented")
+}
+
+func (sample *TDigestSample) Values() []int64 {
+	panic("not implemented")
+}
+
+func (sample *TDigestSample) merged() *bucket {
+	if len(sample.buckets) == 0 {
+		return sample.cur
+	}
+
+	ret := sample.newBucket()
+	for _, b := range sample.buckets {
+		ret.merge(b)
+	}
+	ret.merge(sample.cur)
+	return ret
+}
+
+func (sample *TDigestSample) Max() int64 {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	merged := sample.merged()
+	if merged.count == 0 {
+		return 0
+	}
+	return int64(merged.digest.Max())
+}
+
+func (sample *TDigestSample) Mean() float64 {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	merged := sample.merged()
+	if merged.count == 0 {
+		return 0.0
+	}
+	return float64(merged.values) / float64(merged.count)
+}
+
+func (sample *TDigestSample) Min() int64 {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	merged := sample.merged()
+	if merged.count == 0 {
+		return 0
+	}
+	return int64(merged.digest.Min())
+}
+
+func (sample *TDigestSample) Percentile(percentile float64) float64 {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	merged := sample.merged()
+	if merged.count == 0 {
+		return 0.0
+	}
+	return merged.digest.Quantile(percentile)
+}
+
+func (sample *TDigestSample) Percentiles(percentiles []float64) []float64 {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	merged := sample.merged()
+	ret := make([]float64, 0, len(percentiles))
+	for _, p := range percentiles {
+		ret = append(ret, merged.digest.Quantile(p))
+	}
+	return ret
+}
+
+func (sample *TDigestSample) Snapshot() _metrics.Sample {
+	sample.mutex.RLock()
+	defer sample.mutex.RUnlock()
+
+	sample.drop()
+	merged := sample.merged()
+	cp := *merged.digest
+	return &TDigestSample{
+		timeWindow: sample.timeWindow,
+		count:      sample.count,
+		cur: &bucket{
+			earliest: merged.earliest,
+			count:    merged.count,
+			values:   merged.values,
+			digest:   &cp,
```

Comment:
there was a method to create merging data from another digest which doesn't exist in the tdigest version. I think a simple copy should suffice here as all the data in the struct is actual copyable concretes.

Reply:
@i0exception